Fix race in storage join

vdimir opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Block parallel insertions into storage join

Fix race introduced in PR #20079

vdimir committed:
Try to fix race in storage join: block parralel inserts

vdimir committed:
Add test race_condition_storage.sh

vdimir committed:
Lock before reuseJoinedData

vdimir commented:
`test_storage_kafka/test.py::test_kafka_no_holes_when_write_suffix_failed` not related, failed in master during past three days

vdimir committed:
Remove unused *Locked methods from HashJoin, add comments

vdimir committed:
Rename 01731_race_condition_storage.sh -> 01732_race_condition_storage_long

KochetovNicolai commented:
Some considerations:
* `std::shared_lock<std::shared_mutex> storage_join_lock` is not needed in `HashJoin`. `HashJoin` should not be responsible for external synchronization.
* `JoinSource` is an internal class for `HashJoin`. It need to get `HashJoinPtr` +  `std::shared_lock<std::shared_mutex> storage_join_lock` in constructor (btw, why this lock was removed?)
* Function `JoinGet` does not need to have access to `HashJoinPtr`. We can add `StorageJoin::joinGet`, which takes `std::shared_lock<std::shared_mutex> storage_join_lock` and calls `HashJoin::joinGet`. Lock will be taken for each block, but it is ok (and we will have less chances to have a deadlock).

Also, IMHO, `joinGet` function should be a member of `HashJoin::RightTableData` (but this is not important for this pr).

KochetovNicolai commented on src/Storages/StorageJoin.h:
```diff
@@ -31,12 +43,9 @@ class StorageJoin final : public ext::shared_ptr_helper<StorageJoin>, public Sto
     /// Access the innards.
-    HashJoinPtr & getJoin() { return join; }
+    std::shared_ptr<HashJoinHolder> getJoin() { return std::make_shared<HashJoinHolder>(rwlock, join); }
     HashJoinPtr getJoin(std::shared_ptr<TableJoin> analyzed_join) const;
```
I think this method should actually return `HashJoinHolder` (looks a little bit more logical then having `shared_lock` inside `HashJoin`) .

vdimir replied:
We discussed that it will requires changes in [ExpressionAnalyzer](https://github.com/ClickHouse/ClickHouse/blob/36c178c0231a073fdb3f75ca52a85ba9e4fbea7e/src/Interpreters/ExpressionAnalyzer.cpp#L808), it will have to handle this lock. So decided to keep lock in HashJoin.

vdimir committed:
Update StorageJoin locking

Move joinGet into StorageJoin

Protect JoinSource with lock, add test

Add comments about locking logic

vdimir commented:
> `std::shared_lock<std::shared_mutex> storage_join_lock` is not needed in `HashJoin`. `HashJoin` should not be responsible for external synchronization.

We discussed that it will requires changes in ExpressionAnalyzer, it will have to handle this lock. So decided to keep lock in HashJoin and add mote detailed comments to code.

> `JoinSource` is an internal class for `HashJoin`. It need to get `HashJoinPtr` +  `std::shared_lock<std::shared_mutex> storage_join_lock` in constructor (btw, why this lock was removed?)

Yes, lock removed accidentally, I fixed it and added into test.
 
> Function `JoinGet` does not need to have access to `HashJoinPtr`. We can add `StorageJoin::joinGet`, which takes `std::shared_lock<std::shared_mutex> storage_join_lock` and calls `HashJoin::joinGet`. Lock will be taken for each block, but it is ok (and we will have less chances to have a deadlock).

I moved `joinGet` into `StorageJoin`, and take lock for each block, because `used_flags` are not required in this method.

vdimir committed:
Add static_assert to joinGet to keep information about current locking logic

KochetovNicolai commented on src/Functions/FunctionJoinGet.h:
```diff
@@ -9,14 +9,15 @@ namespace DB
 public:
-    ExecutableFunctionJoinGet(HashJoinPtr join_, const DB::Block & result_columns_)
-        : join(std::move(join_)), result_columns(result_columns_) {}
+    ExecutableFunctionJoinGet(StorageJoinPtr storage_join_, const DB::Block & result_columns_)
```
I think we also need to hold `TableLockHolder` here along with storage.

KochetovNicolai commented on src/Interpreters/HashJoin.cpp:
```diff
@@ -652,7 +639,7 @@ bool HashJoin::addJoinedBlock(const Block & source_block, bool check_limits)
 
     {
-        std::unique_lock lock(data->rwlock);
+        assert(storage_join_lock.mutex() == nullptr);
```
Hm, I think exception will be better here, cause it is not a class invariant.

KochetovNicolai commented on src/Storages/StorageJoin.h:
```diff
@@ -30,12 +29,12 @@ class StorageJoin final : public ext::shared_ptr_helper<StorageJoin>, public Sto
 
-    /// Verify that the data structure is suitable for implementing this type of JOIN.
-    void assertCompatible(ASTTableJoin::Kind kind_, ASTTableJoin::Strictness strictness_) const;
+    DataTypePtr joinGetCheckAndGetReturnType(const DataTypes & data_types, const String & column_name, bool or_null) const;
+    ColumnWithTypeAndName joinGet(const Block & block, const Block & block_with_columns_to_add) const;
```
Need some comments for this methods.

vdimir committed:
Minor changes in Hash/StorageJoin, add TableLockHolder to ExecutableFunctionJoinGet

vdimir committed:
Merge remote-tracking branch 'upstream/master' into fix-race-storage-join
